spatialdata/_core/transform.py

```diff
@@ -1,86 +1,526 @@
 
 
-@get_transform.register
-def _(arg: np.ndarray) -> Transform:  # type: ignore[type-arg]
-    return Transform(
-        ndim=arg.ndim,
-    )
+class Coordinates(BaseTransformation):
+    def __init__(self) -> None:
+        raise NotImplementedError()
 
+    # @property
+    # def ndim(self) -> Optional[int]:
+    #     return self._ndim
 
-@get_transform.register
-def _(arg: ad.AnnData) -> Transform:
-    # check if the AnnData has transform information, otherwise fetches it from default scanpy storage
-    if "transform" in arg.uns:
-        return Transform(
-            translation=arg.uns["transform"]["translation"], scale_factors=arg.uns["transform"]["scale_factors"]
-        )
-    elif "spatial" in arg.uns and "spatial" in arg.obsm:
-        ndim = arg.obsm["spatial"].shape[1]
-        libraries = arg.uns["spatial"]
-        assert len(libraries) == 1
-        library = libraries.__iter__().__next__()
-        scale_factors = library["scalefactors"]["tissue_hires_scalef"]
-        return Transform(
-            translation=np.zeros(ndim, dtype=float), scale_factors=np.array([scale_factors] * ndim, dtype=float)
-        )
-    else:
-        return Transform(ndim=2)
+
+class InverseOf(BaseTransformation):
+    def __init__(self, transformation: Union[Dict[str, Any], BaseTransformation]) -> None:
+        if isinstance(transformation, BaseTransformation):
+            self.transformation = transformation
+        else:
+            self.transformation = get_transformation_from_dict(transformation)
+        self._ndim = self.transformation.ndim
+
+    @property
+    def ndim(self) -> Optional[int]:
+        return self._ndim
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "type": "inverseOf",
+            "transformation": self.transformation.to_dict(),
+        }
+
+    def transform_points(self, points: ArrayLike) -> ArrayLike:
+        return self.transformation.inverse().transform_points(points)
+
+    def inverse(self) -> BaseTransformation:
+        return self.transformation
+
+
+class Bijection(BaseTransformation):
+    def __init__(
+        self, forward: Union[Dict[str, Any], BaseTransformation], inverse: Union[Dict[str, Any], BaseTransformation]
+    ) -> None:
+        if isinstance(forward, BaseTransformation):
+            self.forward = forward
+        else:
+            self.forward = get_transformation_from_dict(forward)
+
+        if isinstance(inverse, BaseTransformation):
+            self._inverse = inverse
+        else:
+            self._inverse = get_transformation_from_dict(inverse)
+        assert self.forward.ndim == self._inverse.ndim
+        self._ndim = self.forward.ndim
+
+    @property
+    def ndim(self) -> Optional[int]:
+        return self._ndim
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "type": "bijection",
+            "forward": self.forward.to_dict(),
+            "inverse": self._inverse.to_dict(),
+        }
```

Comment:
I feel we might want to put transform returns somewhere else (maybe a dataclass) ? and only populate them in the class method call.

Reply:
Yes, I agree we should separate the two things.

Reply:
I didn't use dataclasses but I went toward what you said by doing the following.
The conversion between how things are stored on disc (json representation of python dicts with native types) and how are stored in memory (transformation and coordinate system classes) is now done only in the functions `from_dict()` and `to_dict()`. Before it was messy because the constructor of each transformation class could take as input both the dict representation or explicit constructor parameters (example: before a `Sequence` could be defined by a list of `BaseTransformation` or by a list of `Dict`, now only by a list of `Transformation`; another example: before a `Affine` could take as input either the matrix representation of the affine transformation, either the list of list of float specified by the NGFF specs, which is basically the matrix without the last row. Now only the first matrix representation is allowed).

I don't like that `from_dict()` is in `BaseTransformation` and has a long list of `if else`, while `to_dict()` is specified in each class, we could make this uniform.